### PR TITLE
Bugfix/use write head instead of implicit header

### PR DESCRIPTION
**Pull request by Icehunter**
updated PR with rebase from express/compress:master to handle #128

### CODE REVIEW

**Comment by dougwilson** [comment]
Just wanted to note on here it looks like just a rebase, without addressing the review comments yet. I presume that will come in follow up commits?

**Comment by Icehunter** [comment]
> Just wanted to note on here it looks like just a rebase, without addressing the review comments yet. I presume that will come in follow up commits?

Yes, working on that now.

**Comment by Icehunter** [comment]
@dougwilson I found all the comments I think needed to be fixed, except 2 perhaps that I'll work on tonight. 

1. the on('data') callback which I think doesn't need a data check; per the noop it's just a way to prevent hangs
2. the comments from @sogaani regarding the response code. I'll look at that as well.

**Comment by Icehunter** [comment]
@dougwilson Is there anything more you see for this PR to have?
@sogaani with regards to your comment on the error status code; it seems to be a big edge case. I work on a very large API and haven't encountered this issue before.

**Comment by dougwilson** [comment]
I was under the impression you were going to make changes in regards to the two bullet points above, which is what I'm waiting on. I never saw any additional comments after that message. Are you still working on them?

**Comment by Icehunter** [comment]
Eh! Sorry; I did leave the on('data') callback as is in the test, but I removed all the logs and I believe covered everything you had commented on in the previous test.

For the second bullet point I quite literally have no clue off the top of my head how to even cover that in an edge case test.

**Comment by valoricDe** [comment]
Seems like this one got stuck right?

**Comment by lamweili** [comment]
For the clarity that #129 is not in this PR, the PR description should be changed to:
> updated PR with rebase from express/compress:master to handle #128 ~~and #129~~

File: test/compression.js [hunk 1]
```diff
@@ -305,6 +315,31 @@ describe('compression()', function () {
...
+        res.end('hello, world')
+      })
+      server.on('listening', function () {
+        var client = createHttp2Client(server.address().port)
+        var request = client.request({
+          'Accept-Encoding': 'gzip'
+        })
+        request.on('response', function (headers) {
+          assert.strictEqual(headers['content-encoding'], 'gzip')
+        })
```
Comment (lamweili) [inline test/compression.js:331]:
```suggestion
        })
        var chunks = []
```

File: test/compression.js [hunk 2]
```diff
@@ -305,6 +315,31 @@ describe('compression()', function () {
...
+        request.on('data', function (chunk) {
+          // no-op without which the request will stay open and cause a test timeout
```
Comment (lamweili) [inline test/compression.js:333]:
```suggestion
          chunks.push(chunk);
```

File: test/compression.js [hunk 3]
```diff
@@ -305,6 +315,31 @@ describe('compression()', function () {
...
+        })
+        request.on('end', function () {
+          closeHttp2(request, client, server, done)
```
Comment (lamweili) [inline test/compression.js:336]:
```suggestion
          zlib.gunzip(Buffer.concat(chunks), function(err, data) { 
            assert.strictEqual(data.toString(), 'hello, world')
            closeHttp2(client, server, done)
          })
```

File: test/compression.js [hunk 4]
```diff
@@ -674,6 +709,51 @@ function createServer (opts, fn) {
...
+  })
+  server.listen(0, '127.0.0.1')
+  return server
+}
+
+function createHttp2Client (port) {
+  return http2.connect('http://127.0.0.1:' + port)
+}
+
+function closeHttp2 (request, client, server, callback) {
```
Comment (lamweili) [inline test/compression.js:733]:
```suggestion
function closeHttp2 (client, server, callback) {
```

File: test/compression.js (hunk 1 above)
Comment (lamweili) [inline test/compression.js:331]:
```suggestion
        })
        var chunks = [];
```

File: test/compression.js (hunk 2 above)
Comment (lamweili) [inline test/compression.js:333]:
```suggestion
          chunks.push(chunk);
```

File: test/compression.js (hunk 3 above)
Comment (lamweili) [inline test/compression.js:336]:
```suggestion
          zlib.gunzip(Buffer.concat(chunks), function(err, data) { 
            assert.strictEqual(data.toString(), 'hello, world')
            closeHttp2(client, server, done)
          })
```

File: test/compression.js (hunk 4 above)
Comment (lamweili) [inline test/compression.js:733]:
```suggestion
function closeHttp2 (client, server, callback) {
```

File: test/compression.js [hunk 5]
```diff
@@ -674,6 +709,51 @@ function createServer (opts, fn) {
...
+  if (typeof client.shutdown === 'function') {
+    // this is the node v8.x way of closing the connections
+    request.destroy(http2.constants.NGHTTP2_NO_ERROR, function () {
```
Comment (lamweili) [inline test/compression.js:736]:
```suggestion
```

File: test/compression.js [hunk 6]
```diff
@@ -674,6 +709,51 @@ function createServer (opts, fn) {
...
+      client.shutdown({}, function () {
+        server.close(function () {
+          callback()
+        })
+      })
+    })
```
Comment (lamweili) [inline test/compression.js:742]:
```suggestion
```

File: test/compression.js [hunk 7]
```diff
@@ -674,6 +709,51 @@ function createServer (opts, fn) {
...
+  } else {
+    // this is the node v9.x onwards way of closing the connections
+    request.close(http2.constants.NGHTTP2_NO_ERROR, function () {
```
Comment (lamweili) [inline test/compression.js:745]:
```suggestion
```

File: test/compression.js [hunk 8]
```diff
@@ -674,6 +709,51 @@ function createServer (opts, fn) {
...
+      client.close(function () {
+        // force existing connections to time out after 1ms.
+        // this is done to force the server to close in some cases where it wouldn't do it otherwise.
+        server.close(function () {
+          callback()
+        })
+      })
+    })
```
Comment (lamweili) [inline test/compression.js:753]:
```suggestion
```

**Review by lamweili — CHANGES_REQUESTED** [review]
Hi @Icehunter and @dougwilson,

I addressed the `on('data') callback` as mentioned in https://github.com/expressjs/compression/pull/128#discussion_r368753350 and https://github.com/expressjs/compression/pull/170#issuecomment-593582447 with the following suggestions.
It now asserts that the data is correct. 

After accepting the suggestions, you might want to correct the indentations within the `function closeHttp2` at line 733.

I believe with this, all outstanding review comments in #128 are addressed and this PR should supersede #128.

**Comment by lamweili** [comment]
@Icehunter, you might want to correct the indentations at line 740-744 and line 747-753.
I didn't include them previously so that https://github.com/expressjs/compression/pull/170/commits/38fe9fd5756f524e438c9abb9e91e7193921e0e3 is surgical. 


https://github.com/expressjs/compression/blob/38fe9fd5756f524e438c9abb9e91e7193921e0e3/test/compression.js#L737-L756

**Comment by lamweili** [comment]
@Icehunter, for the clarity that #129 is not in this PR, the description should change:
> updated PR with rebase from express/compress:master to handle #128 ~~and #129~~

**Comment by Icehunter** [comment]
> @Icehunter, for the clarity that #129 is not in this PR, the description should change:
> 
> > updated PR with rebase from express/compress:master to handle #128 ~and #129~

Updated and rebased again :)

**Comment by lamweili** [comment]
@dougwilson When you are free, can you review this PR?

**Comment by lamweili** [comment]
@dougwilson When you are free, can you review this PR?
Fixes #122 and supersedes #128.
Many thanks! 🤗

I didn't have access in the [forked repository](https://github.com/Icehunter/compression/tree/bugfix/use_writeHead_instead_of_implicitHeader) to run the GH actions. So I forked the PR.
I ran the GH actions on my own repository and here is the result for the latest commit (4c1d27b):
https://github.com/lamweili/compression/actions

**Comment by lamweili** [comment]
@dougwilson, have you got the time to take a look at this PR?

**Comment by lamweili** [comment]
@dougwilson, have you got the time to take a look at this PR?

**Comment by lamweili** [comment]
@dougwilson, have you got the time to take a look at this PR?

File: test/compression.js [hunk 9]
```diff
@@ -7,6 +7,16 @@ var http = require('http')
...
 var zlib = require('zlib')
 
+var describeHttp2 = describe.skip
+try {
+  var http2 = require('http2')
+  describeHttp2 = describe
+} catch (err) {
+  if (err) {
+    console.log('http2 tests disabled.')
+  }
```
Comment (bjohansebas) [inline test/compression.js:17]: Why would we send a message to the console? Wouldn't it be enough to just ignore it?

Comment (lamweili) [reply test/compression.js:17]: Made it verbose previously. Yes, we can ignore it, since the test results will mention it as skipped anyway.

**Review by bjohansebas — APPROVED** [review]

File: test/compression.js (hunk 9 above)
Comment (bjohansebas) [reply test/compression.js:17]: I’m not going to block this PR just because of a message in the console that will be ignored anyway.

**Review by lamweili — COMMENTED** [review]
LGTM

(can't approve cause I don't have explicit access to this repository)
😮‍💨

**Comment by bjohansebas** [comment]
sorry @Icehunter, Can you fix the code style?

**Comment by Icehunter** [comment]
> sorry @Icehunter, Can you fix the code style?

Sure thing.

**Comment by jonchurch** [comment]
Dropping some context from slack re: this PR

> The original issue here is compression relying on node http internals since the days of connect in 2011
> 
> Here's when the commit w/ `_implicitHeader` originally landed:
> 
> https://github.com/senchalabs/connect/commit/715e4f56f8e4ad75baa35a7ae049bd69ca8bf6c6
> 
> http2 module has a different API and internals than http1 in node. So that method doesn't exist at all in http2, throwing when used.
> 
> Compression module doesn't support http2 explicitly. It likely can, but was never intended to.
> `writeHead` has been around since that first commit, but for reasons lost to time wasn't used

Idk that I can say that issue #122 "supporting http2" is closed after landing #170, because I personally haven't reviewed the code or tested for http2 edgecases.

However, removing an http internal to replace w/ a public interface which is equivalent is 💯

**Review by jonchurch — APPROVED** [review]

**Comment by jonchurch** [comment]
This history is long w/ 19 commits, @bjohansebas when/if you merge I'd suggest squashing via the dropdown to keep the history clean
